An unprojected SECOND dimension draws a grouped chart with no series and says nothing — `hasNoCategoryKey` has no counterpart on the series axis

Measured while implementing #4673 (PR #4681), filed rather than widened into it.

## What is wrong

The "known to be empty versus cannot know" split is now answered on both of the pivot's dimensions for the BUCKETING half (#4673): a row whose second-dimension value is `null`/`''` gets a drawn series, and a row that does not carry the group key at all gets no bucket and contributes no column.

The ANNOUNCING half only exists for the first dimension. `AdvancedChartImpl`'s `hasNoCategoryKey` (framework#4033) names an unprojected x-axis dimension instead of drawing an empty axis. There is no counterpart for the series dimension, so a pivot whose second dimension was never projected renders as:

- axes, grid, tooltip and legend frame drawn;
- zero series, therefore zero marks;
- no explanation anywhere.

Concretely: `dimensions: ['status', 'priority']`, `values: ['est_hours']`, where the result rows carry `status` and `est_hours` but no `priority` key at all. `buildChartSeries` returns `series: []` and one bucket row per status.

## Why this is worth a card rather than a shrug

It is the same sentence #4466 was filed over — a chart that is not merely empty but says nothing about why — and the platform has already ruled twice that "cannot know" must REFUSE LOUDLY rather than draw nothing (framework#4033, and the delegated ruling on #4673, 2026-08-15). An empty grouped chart currently reads to the author as "no data matched", when what happened is that a dimension they grouped by never arrived.

Note this behaviour is not NEW in #4673: before it, the unprojected group's measure was written under the `''` key, which was also bound to no series, so the mark count was the same. What changed is only that the phantom column is no longer written.

## Not the same card as #4507

#4507 is that the pivot writes `[xKey]` onto every bucket it creates, so `hasNoCategoryKey` can never fire for the FIRST dimension of a pivoted chart — a guard that exists but is blinded. This is the SECOND dimension, where no guard exists to blind. Fixing #4507 does not produce this one's answer.

Related: #4507, #4673, #4681, #4466, framework#4033.

---
_Generated by [Claude Code](https://claude.ai/code/session_01RnQd8iMMUwXQEV1crFmQiQ)_

<!-- os-dev-report -->
```json
{
  "issue": 4683,
  "status": "done",
  "branch": "claude/issue-4683-series-axis-refusal",
  "pr": "https://github.com/objectstack-ai/objectui/pull/4696",
  "premise_still_valid": true,
  "summary": "Re-derived premise A post-#4681 rather than trusting the card: buildChartSeries(rows, ['status','priority'], ['est_hours']) with priority unprojected returns {data:[{status:'Backlog'},{status:'Done'}], xAxisKey:'status', series:[]}, and the renderer drew an svg with 0 marks and no message. AdvancedChartImpl now refuses that shape with a named placeholder (data-chart-error='no-plottable-series') plus the same console.warn diagnostic pair the framework#4033 guard prints, mirroring its mechanism and its copy convention exactly (hardcoded English in JSX, no locale key - measured before writing, so no locale-pack key was added). The signal is buildChartSeries' own output (series is an array that came out empty, with rows present), never groupKey in row - that is #4507's trap inverted, since the renderer is handed the PIVOTED bucket rows whose columns are the group's VALUES, so a key-in-row test would refuse every grouped chart in the product; #4507's own blindness is untouched. The refusal is limited to the families whose marks come only from series.map (bar, horizontal-bar, line, area, combo) - measured in-file: pie/donut/funnel/radar/scatter/treemap/sankey fall back to series[0]?.dataKey || 'value' and draw a working chart with no series at all. Two deliberate honesty boundaries, both surfaced rather than papered over: the message does NOT name the group key (see open_questions), and series === undefined is left byte-for-byte alone (filed as #4695). Premise C measured rather than guessed: hasNoCategoryKey refuses only when NOT ONE row carries the key, so partial projection draws what projects - the series axis inherits that line for free and it is pinned with its emitted rows. #4694's click-handler region was not edited; core chart-series.ts was read-only (no helper export turned out to be needed).",
  "tests": "All gates run AFTER the final commit, at HEAD bf0d7351e with a clean tree (git status --porcelain empty). (1) pnpm exec vitest run packages/core/ packages/plugin-charts/ -> 'Test Files 117 passed (117) / Tests 2073 passed (2073)'. (2) New file alone (AdvancedChartImpl.unprojectedSeriesDimension.test.tsx, 11 cases) -> 'Test Files 1 passed (1) / Tests 11 passed (11)'. (3) Downstream consumer sweep, run by package name (plugin-dashboard consumes plugin-charts via ChartRenderer): pnpm exec vitest run packages/plugin-dashboard/ -> 'Test Files 58 passed (58) / Tests 462 passed (462)'. (4) Build closure first, then type-check: pnpm --filter '@object-ui/plugin-charts^...' build then pnpm --filter @object-ui/plugin-charts type-check -> exit 0. NOTE the zero-match trap fired for real here: the package has no 'typecheck' script, only 'type-check' - pnpm --filter ... typecheck returned ERR_PNPM_RECURSIVE_RUN_NO_SCRIPT, which would have read as a pass if not checked. (5) pnpm exec eslint --quiet on both changed files -> clean. (6) node scripts/check-changeset-presence.mjs -> OK, '2 source file(s) of 1 released package(s) changed, and this change declares 1 changeset(s)'. (7) pnpm run check:control-bytes -> OK (4225 tracked text files). (8) Re-derived from changed paths beyond the prompt's list: check:i18n-keys, check:i18n-drift (both green, and load-bearing here since the guard adds user-visible copy - confirms the hardcoded-string convention matches the model and drifts no en value), check:phantom-deps (green; no import added). check:spec-symbols / check:action-forward-parity / check:skills-paths are not implicated by a renderer guard + test + changeset. REVERSE VERIFICATION, direction predicted before running: ablate the guard from the committed state (git checkout origin/main -- AdvancedChartImpl.tsx) and exactly the three refusal-arm cases go red while the eight draw/unchanged pins stay green - including 'yields to the category-axis refusal', which is green either way because it asserts the OTHER guard fires. Observed exactly that: '3 failed | 13 passed', the failures being 'names the failure instead of drawing a frame with no marks', 'warns once...', 'fires on every family whose marks come only from series'. Restored via git checkout of the branch (never git stash), verified byte-identical (git status --porcelain empty, git diff HEAD empty), re-ran green. FLAKE, measured on BOTH sides so it is not this change: the first plugin-dashboard sweep had 1 failure at DatasetWidget.dottedDimension.test.tsx:269 ('expected [] to deeply equal [ crm_opportunity ]'); with the guard ablated the full suite was 58/58 green, with it restored 58/58 green, and the file alone is green. It is the second site of the race already filed as #4487 - attached there, not re-filed. PR labels read back after the bots settled: plugin, plugin: charts, tests. No skip-changeset label applies (this PR ships a real changeset and the presence gate requires one).",
  "open_questions": [
    {
      "question": "The refusal names the failure and both authoring causes, but NOT the missing dimension key - unlike hasNoCategoryKey, which names its key. Measured reason: the renderer is never told which dimension the pivot grouped by, and the two upstream shapes reach it byte-identically - buildChartSeries(rows,['status','priority'],['est_hours']) with priority unprojected and buildChartSeries(rows,['status'],[]) with no measure selected BOTH return {data:[{status:'Backlog'},{status:'Done'}],xAxisKey:'status',series:[]}. Should a follow-up plumb the pivoted dimension into the renderer so the message can name it?",
      "options": [
        "A - ship as landed: renderer-local signal, copy names the failure plus both causes, console.warn carries the diagnostic pair (axis key + the keys the rows actually carry). 1 file, fires on every path that can produce the harm, including future ones.",
        "B - plumb it: export a pure pivotSeriesDimension(dimensions, values) from core (the pivot rule stated once), have ObjectChart and DatasetWidget pass it, ChartRenderer forward it, and name the key in the copy. Touches 5 files across 3 packages (core, plugin-charts, plugin-dashboard).",
        "C - B, folded together with #4695 (the series === undefined arm), since one plumbing round makes both reachable at once."
      ],
      "recommendation": "A now, C as the follow-up if the maintainer wants key-level naming. Real business need: the author's actual need is to stop reading an empty frame as 'no data matched', which A fully serves - the warn already prints the axis key and the row keys, and if the chart has a measure selected the author knows immediately which cause it is; B buys one noun in one sentence. Long-term soundness for this project: B introduces a prop that every present AND future chart call site must remember to pass, and a call site that forgets loses the guard silently - which is precisely the blindable-guard class #4507 exists to record, so B trades a guard that cannot be blinded for one that can. A's signal is intrinsic to the renderer (marks come only from series), so it closes the class rather than two known call paths. Hard to get wrong for AI-authored metadata: A fires on every entry path including DatasetWidget's, which does not carry dimensions on its chart schema at all and would need its own plumbing under B - so under B a dashboard dataset widget would keep saying nothing until someone noticed. The three axes agree; they only diverge on copy quality, which is the cheaper thing to improve later. Recorded rather than decided unilaterally because it is a contract-shaped question (a new renderer prop and a new core export), and the card ruled on mechanism, not on plumbing."
    }
  ],
  "out_of_scope_findings": [
    "filed as #4695: a cartesian chart handed NO series binding at all (series === undefined, never went through buildChartSeries) draws the same silent empty frame this card just fixed for series: []. Unassigned, 'finding' label (observation-class: pre-existing, no data understated). Left out deliberately - widening the guard there changes behaviour for every hand-authored chart, which is an author-contract decision, not a mechanism fix. Current behaviour is PINNED by a test in this PR so it cannot later read as a silent contradiction.",
    "attached to #4487 (comment 5301827655) rather than re-filed: DatasetWidget.dottedDimension.test.tsx:269 is a second site of the installMetaRouter race that card already describes at dottedDimensionTable.test.tsx:207 - identical shape (waitFor on the rendered cell, then a bare assertion on 'requested'), identical failure text. Searched first; the prior filing is why no duplicate exists. Measured green on both sides of this change."
  ]
}
```


---
_Generated by [Claude Code](https://claude.ai/code)_

--- a/.changeset/chart-unprojected-series-dimension.md
+++ b/.changeset/chart-unprojected-series-dimension.md
@@ -1,0 +1,28 @@
+---
+'@object-ui/plugin-charts': patch
+---
+
+A grouped chart whose SECOND dimension was never projected says so instead of
+drawing an empty frame.
+
+"Cannot know refuses loudly" was answered on the first dimension only.
+`AdvancedChartImpl`'s `hasNoCategoryKey` (framework#4033) names an unprojected
+x-axis dimension rather than drawing a bare axis; the series axis had no
+counterpart, so a pivot whose second dimension was absent from the result rows
+produced `series: []` and rendered axes, grid, tooltip and legend around zero
+marks — indistinguishable, to the author, from "no data matched".
+
+Such a chart now renders the same explanatory placeholder
+(`data-chart-error="no-plottable-series"`) and logs the same diagnostic pair the
+category-axis guard logs: the axis it did plot, and the keys its rows actually
+carry.
+
+The three-way distinction is unchanged and pinned: null / empty-string group
+values still DRAW (they are real groups with real buckets), a partially
+projected group key still draws what projects — mirroring the category axis,
+which refuses only when NOT ONE row carries the key — and an ordinary pivot
+renders unchanged. The refusal is limited to the families whose marks come from
+`series` and nothing else (bar, horizontal-bar, line, area, combo); pie, donut,
+funnel, radar and scatter draw from a `value` column with no series declared, so
+they are untouched. A caller that computed no series binding at all
+(`series === undefined`) is also untouched.

--- a/packages/plugin-charts/src/AdvancedChartImpl.tsx
+++ b/packages/plugin-charts/src/AdvancedChartImpl.tsx
@@ -1344,6 +1344,116 @@ function hasNoCategoryKey(props: AdvancedChartImplProps): boolean {
 }
 
 /**
+ * Chart families whose marks come from `series` AND NOTHING ELSE — the set the
+ * series-axis guard below is allowed to refuse for.
+ *
+ * MEASURED in this file, not assumed, and deliberately NARROWER than
+ * {@link CATEGORY_AXIS_CHART_TYPES}: `bar` / `horizontal-bar` / `line` / `area`
+ * share one cartesian tail that renders `series.map(…)` and nothing else, and
+ * `combo`'s `ComposedChart` does the same. Every other family reads
+ * `series[0]?.dataKey || 'value'` — pie, donut, funnel, radar, scatter,
+ * treemap, sankey all fall back to a `value` column and draw a perfectly good
+ * chart with no series declared at all, so a refusal there would blank a
+ * WORKING chart rather than explain a broken one.
+ */
+const SERIES_ONLY_CHART_TYPES: ReadonlySet<string> = new Set([
+  'bar', 'horizontal-bar', 'line', 'area', 'combo',
+]);
+
+/**
+ * The series-axis half of the same doctrine — objectui#4683.
+ *
+ * `hasNoCategoryKey` above answers "cannot know refuses loudly" for the FIRST
+ * dimension. A pivoted chart has a second one, and it had no answer at all: a
+ * pivot whose SECOND dimension was never projected produces
+ * `series: []` with one bucket row per category, so the renderer drew axes,
+ * grid, tooltip and legend around ZERO marks and said nothing. To the author
+ * that reads as "no data matched", when what happened is that a dimension they
+ * grouped BY never arrived.
+ *
+ * ## What this reads, and what it deliberately does NOT read
+ *
+ * The signal is `series: []` — {@link buildChartSeries}' own output — plus rows
+ * to draw it against. NOT `groupKey in row`, which is the objectui#4507 trap
+ * INVERTED: this component is handed the PIVOTED bucket rows, whose columns are
+ * the group's VALUES, so the second dimension's own key is a column of no pivot
+ * ever, ordinary ones included. A `key in row` test here would refuse every
+ * grouped chart in the product.
+ *
+ * That keeps the three-way distinction intact, and each arm is pinned by a test:
+ *
+ *   - **null / `''` group values DRAW** (objectui#4673) — they are real groups,
+ *     they get real buckets, so `series` is non-empty and this never fires;
+ *   - **an unprojected group key REFUSES** — no row carries it, so no bucket
+ *     exists, so `series` is `[]`: this fires;
+ *   - **an ordinary pivot renders unchanged**, and so does a PARTIALLY
+ *     projected one. That last is the mirror of `hasNoCategoryKey`'s own
+ *     `!rows.some(…)`: the first dimension refuses only when NOT ONE row
+ *     carries the key, and draws what projects otherwise. A pivot where some
+ *     rows carry the group key yields those groups' series, so this stays
+ *     silent and the chart draws what projected.
+ *
+ * ## Why the message cannot name the group key, and why that is honest
+ *
+ * `hasNoCategoryKey` names the key it could not find. This cannot, because the
+ * renderer is not told which dimension the pivot grouped by — and the two
+ * upstream shapes that reach it are byte-identical. Measured:
+ * `buildChartSeries(rows, ['status','priority'], ['est_hours'])` with `priority`
+ * unprojected and `buildChartSeries(rows, ['status'], [])` with no measure
+ * selected BOTH return `{data:[{status:'Backlog'},…], xAxisKey:'status',
+ * series:[]}`. Naming one cause would be a sentence that is false half the time,
+ * which is worse than the silence this replaces — so the copy names the failure
+ * and BOTH authoring causes, and the console warning carries the diagnostic pair
+ * (`xAxisKey` + the keys the rows actually carry) exactly as the model does.
+ *
+ * ## Why `[]` and not "no series at all"
+ *
+ * `Array.isArray(series) && series.length === 0` — a binding that was COMPUTED
+ * and came out empty, which is what both `buildChartSeries` call paths hand
+ * over. `series === undefined` means no binding was ever computed (a caller that
+ * never went through the helper); those charts are left byte-for-byte as they
+ * were.
+ */
+function hasNoPlottableSeries(props: AdvancedChartImplProps): boolean {
+  const chartType = props.chartType === 'column' ? 'bar' : (props.chartType ?? 'bar');
+  const rows = Array.isArray(props.data) ? props.data : [];
+  return (
+    SERIES_ONLY_CHART_TYPES.has(chartType) &&
+    rows.length > 0 &&
+    Array.isArray(props.series) &&
+    props.series.length === 0
+  );
+}
+
+/**
+ * The shell both refusals render — one placeholder, two diagnoses.
+ *
+ * Stated once so the series-axis guard cannot drift from the framework#4033 one
+ * it mirrors: same box, same `role="status"` (a refusal is a state, not an
+ * alert), same muted centred type, and a `data-chart-error` code naming WHICH
+ * refusal fired.
+ */
+function ChartRefusal({
+  code,
+  className,
+  children,
+}: {
+  code: string;
+  className?: string;
+  children: React.ReactNode;
+}) {
+  return (
+    <div
+      className={`flex h-full min-h-[120px] w-full items-center justify-center p-4 text-center text-sm text-muted-foreground ${className ?? ''}`}
+      role="status"
+      data-chart-error={code}
+    >
+      <span>{children}</span>
+    </div>
+  );
+}
+
+/**
  * Public entry point. A THIN wrapper purely so the unreadable-data guard can
  * short-circuit without breaking the rules of hooks: the condition depends on
  * `data`, which arrives asynchronously, so an early return inside the renderer
@@ -1352,6 +1462,11 @@ function hasNoCategoryKey(props: AdvancedChartImplProps): boolean {
  */
 export default function AdvancedChartImpl(props: AdvancedChartImplProps) {
   const missingCategoryKey = hasNoCategoryKey(props);
+  // The x-axis refusal WINS when both apply: a chart whose rows carry no
+  // category key at all cannot plot an axis, which is the more fundamental
+  // failure and the more specific message. Without this gate such a chart would
+  // also warn about its series, printing two diagnoses for one cause.
+  const noPlottableSeries = !missingCategoryKey && hasNoPlottableSeries(props);
   const xAxisKey = props.xAxisKey ?? 'name';
   const firstRowKeys = React.useMemo(
     () => Object.keys((Array.isArray(props.data) ? props.data[0] : undefined) ?? {}),
@@ -1371,18 +1486,35 @@ export default function AdvancedChartImpl(props: AdvancedChartImplProps) {
     );
   }, [missingCategoryKey, xAxisKey, firstRowKeys]);
 
+  React.useEffect(() => {
+    if (!noPlottableSeries) return;
+    // The same diagnostic PAIR the guard above prints, for the same reason: the
+    // axis the chart did plot, and the keys its rows actually carry. In the
+    // objectui#4683 shape that second half is the tell — bucket rows carrying
+    // the category and NOTHING else say the group column was never written.
+    console.warn(
+      `[chart] no series to plot against the category axis "${xAxisKey}" — rendering an ` +
+      `explanatory placeholder instead of an empty frame. Row keys: ${JSON.stringify(firstRowKeys)}. ` +
+      `A grouped chart's SECOND dimension must be PROJECTED by the dataset query, and a ` +
+      `chart with no measure selected has nothing to draw either (objectui#4683, framework#4033).`,
+    );
+  }, [noPlottableSeries, xAxisKey, firstRowKeys]);
+
   if (missingCategoryKey) {
     return (
-      <div
-        className={`flex h-full min-h-[120px] w-full items-center justify-center p-4 text-center text-sm text-muted-foreground ${props.className ?? ''}`}
-        role="status"
-        data-chart-error="missing-category-key"
-      >
-        <span>
-          This chart cannot plot its category axis: no row has a{' '}
-          <code className="font-mono">{xAxisKey}</code> field.
-        </span>
-      </div>
+      <ChartRefusal code="missing-category-key" className={props.className}>
+        This chart cannot plot its category axis: no row has a{' '}
+        <code className="font-mono">{xAxisKey}</code> field.
+      </ChartRefusal>
+    );
+  }
+
+  if (noPlottableSeries) {
+    return (
+      <ChartRefusal code="no-plottable-series" className={props.className}>
+        This chart cannot plot any series: no measure or group reached its{' '}
+        <code className="font-mono">{xAxisKey}</code> axis.
+      </ChartRefusal>
     );
   }
 

--- a/packages/plugin-charts/src/AdvancedChartImpl.unprojectedSeriesDimension.test.tsx
+++ b/packages/plugin-charts/src/AdvancedChartImpl.unprojectedSeriesDimension.test.tsx
@@ -1,0 +1,283 @@
+/**
+ * ObjectUI
+ * Copyright (c) 2024-present ObjectStack Inc.
+ *
+ * objectui#4683 — the ANNOUNCING half of "cannot know refuses loudly", on the
+ * SERIES axis.
+ *
+ * `AdvancedChartImpl.missingCategoryKey.test.tsx` pins that answer for the FIRST
+ * dimension (framework#4033): rows that never carried the category key get a
+ * named refusal instead of an axis frame with no marks. A pivot has a SECOND
+ * dimension and had no such answer — an unprojected group key produced
+ * `series: []`, so the chart drew axes, grid, tooltip and legend around ZERO
+ * marks and said nothing. Measured on this file's own fixture with the guard
+ * reverted: an `<svg>`, `0` `.recharts-rectangle` elements, no warning, no text.
+ *
+ * The three-way distinction objectui#4673 landed is what these tests hold still,
+ * one `describe` per arm:
+ *
+ *   - null / `''` group values DRAW — real groups, real buckets, real bars;
+ *   - an unprojected group key REFUSES — nothing exists to draw or to label;
+ *   - an ordinary (and a PARTIALLY projected) pivot renders unchanged.
+ *
+ * Every fixture goes through `buildChartSeries` rather than hand-writing the
+ * props, because the defect lives in the composition the two consumers perform
+ * (`ObjectChart`, `DatasetWidget`) and not in either half alone.
+ */
+
+import React from 'react';
+import { describe, it, expect, afterEach, vi } from 'vitest';
+import { render, cleanup } from '@testing-library/react';
+import { buildChartSeries, NULL_CATEGORY_LABEL } from '@object-ui/core';
+
+// Recharts' ResponsiveContainer measures via ResizeObserver, which reports 0×0
+// under the headless DOM, so nothing paints. Fix its size.
+vi.mock('recharts', async () => {
+  const actual = await vi.importActual<any>('recharts');
+  return {
+    ...actual,
+    ResponsiveContainer: ({ children }: any) =>
+      React.cloneElement(children, { width: 480, height: 320 }),
+  };
+});
+
+import AdvancedChartImpl from './AdvancedChartImpl';
+
+afterEach(cleanup);
+
+const DIMS = ['status', 'priority'];
+const VALS = ['est_hours'];
+
+/**
+ * The card's repro: `dimensions: ['status','priority']`, `values:
+ * ['est_hours']`, and result rows that carry `status` and `est_hours` but no
+ * `priority` KEY AT ALL — grouped by a dimension the query never projected.
+ */
+const UNPROJECTED = [
+  { status: 'Backlog', est_hours: 12 },
+  { status: 'Done', est_hours: 30 },
+];
+
+/** The same pivot, projected. The ordinary case that must not change. */
+const PROJECTED = [
+  { status: 'Backlog', priority: 'High', est_hours: 12 },
+  { status: 'Done', priority: 'Low', est_hours: 30 },
+];
+
+/** Render a pivot result exactly as `DatasetWidget` / `ObjectChart` do. */
+function renderPivot(
+  rows: Array<Record<string, unknown>>,
+  props: Record<string, unknown> = {},
+) {
+  const { data, xAxisKey, series } = buildChartSeries(rows, DIMS, VALS, null, {
+    nullCategoryLabel: NULL_CATEGORY_LABEL,
+  });
+  const { container } = render(
+    <AdvancedChartImpl
+      chartType="bar"
+      data={data as any}
+      xAxisKey={xAxisKey}
+      series={series as any}
+      isAnimationActive={false}
+      {...(props as any)}
+    />,
+  );
+  return { container, data, series };
+}
+
+const refusal = (container: HTMLElement) =>
+  container.querySelector('[data-chart-error="no-plottable-series"]');
+
+describe('AdvancedChartImpl — an unprojected SECOND dimension refuses (objectui#4683)', () => {
+  it('names the failure instead of drawing a frame with no marks', () => {
+    const warn = vi.spyOn(console, 'warn').mockImplementation(() => {});
+    const { container, series } = renderPivot(UNPROJECTED);
+
+    // Premise A, re-derived post-#4681 rather than taken from the card: no row
+    // carries `priority`, so no group has a bucket and no series exists.
+    expect(series).toEqual([]);
+
+    const placeholder = refusal(container);
+    expect(placeholder, 'renders the explanatory placeholder').not.toBeNull();
+    // The axis it DID plot is named, as the first dimension's guard names its
+    // own key — that plus the warning's row keys is the author's diagnosis.
+    expect(placeholder?.textContent).toContain('status');
+    // The old behaviour: axes, grid, tooltip and legend around zero marks.
+    // Anything that still paints a plot here is the silence this replaces.
+    expect(container.querySelector('svg')).toBeNull();
+    expect(container.querySelectorAll('.recharts-rectangle')).toHaveLength(0);
+    warn.mockRestore();
+  });
+
+  it('warns once, naming the axis key and the keys the rows actually carry', () => {
+    const warn = vi.spyOn(console, 'warn').mockImplementation(() => {});
+    renderPivot(UNPROJECTED);
+
+    expect(warn).toHaveBeenCalledTimes(1);
+    const msg = String(warn.mock.calls[0]?.[0] ?? '');
+    expect(msg).toContain('status');
+    // The bucket rows carry the category and NOTHING else — the tell that the
+    // group column was never written. `est_hours` is absent from them because
+    // no group owned it, which is exactly what the author needs to see.
+    expect(msg).toContain('Row keys: ["status"]');
+    expect(msg).toContain('objectui#4683');
+    warn.mockRestore();
+  });
+
+  it('fires on every family whose marks come only from `series`', () => {
+    const warn = vi.spyOn(console, 'warn').mockImplementation(() => {});
+    for (const chartType of ['bar', 'horizontal-bar', 'line', 'area', 'combo', 'column']) {
+      const { container } = renderPivot(UNPROJECTED, { chartType });
+      expect(refusal(container), `${chartType} refuses`).not.toBeNull();
+      cleanup();
+    }
+    warn.mockRestore();
+  });
+
+  it('stays out of the way of families that draw from a `value` column', () => {
+    const warn = vi.spyOn(console, 'warn').mockImplementation(() => {});
+    // pie / donut / funnel / radar / scatter / treemap / sankey all fall back to
+    // `series[0]?.dataKey || 'value'`, so no series is not no chart there — a
+    // refusal would blank a working one.
+    for (const chartType of ['pie', 'donut', 'funnel', 'radar', 'scatter']) {
+      const { container } = render(
+        <AdvancedChartImpl
+          chartType={chartType as any}
+          data={[{ name: 'Backlog', value: 12 }, { name: 'Done', value: 30 }]}
+          xAxisKey="name"
+          series={[]}
+          isAnimationActive={false}
+        />,
+      );
+      expect(refusal(container), `${chartType} draws`).toBeNull();
+      expect(container.querySelector('svg'), `${chartType} paints`).not.toBeNull();
+      cleanup();
+    }
+    expect(warn).not.toHaveBeenCalled();
+    warn.mockRestore();
+  });
+
+  it('leaves a caller that computed NO series binding at all untouched', () => {
+    const warn = vi.spyOn(console, 'warn').mockImplementation(() => {});
+    // `series === undefined` is "no binding was ever computed", not "a binding
+    // came out empty" — a different sentence, and not this card's. Such charts
+    // stay byte-for-byte as they were.
+    const { container } = render(
+      <AdvancedChartImpl
+        chartType="bar"
+        data={[{ status: 'Backlog', est_hours: 12 }]}
+        xAxisKey="status"
+        isAnimationActive={false}
+      />,
+    );
+    expect(refusal(container)).toBeNull();
+    expect(warn).not.toHaveBeenCalled();
+    warn.mockRestore();
+  });
+
+  it('yields to the category-axis refusal when both apply — one cause, one diagnosis', () => {
+    const warn = vi.spyOn(console, 'warn').mockImplementation(() => {});
+    const { container } = render(
+      <AdvancedChartImpl
+        chartType="bar"
+        data={[{ est_hours: 12 }, { est_hours: 30 }]}
+        xAxisKey="status"
+        series={[]}
+        isAnimationActive={false}
+      />,
+    );
+
+    expect(container.querySelector('[data-chart-error="missing-category-key"]')).not.toBeNull();
+    expect(refusal(container)).toBeNull();
+    expect(warn).toHaveBeenCalledTimes(1);
+    expect(String(warn.mock.calls[0]?.[0] ?? '')).toContain('no row has the category key');
+    warn.mockRestore();
+  });
+});
+
+describe('AdvancedChartImpl — what the refusal must NOT swallow (objectui#4683)', () => {
+  it('draws a null / empty-string second-dimension group (objectui#4673 intact)', () => {
+    const warn = vi.spyOn(console, 'warn').mockImplementation(() => {});
+    const { container, series } = renderPivot([
+      { status: 'Backlog', priority: null, est_hours: 12 },
+      { status: 'Done', priority: '', est_hours: 30 },
+    ]);
+
+    // Known to be empty DRAWS: both are real groups with real buckets.
+    expect(series.map((s) => s.dataKey)).toEqual([NULL_CATEGORY_LABEL, '']);
+    expect(refusal(container), 'a known-empty group is not a missing one').toBeNull();
+    expect(container.querySelectorAll('.recharts-rectangle')).toHaveLength(2);
+    expect(warn).not.toHaveBeenCalled();
+    warn.mockRestore();
+  });
+
+  it('renders an ordinary projected pivot unchanged', () => {
+    const warn = vi.spyOn(console, 'warn').mockImplementation(() => {});
+    const { container, series } = renderPivot(PROJECTED);
+
+    expect(series.map((s) => s.dataKey)).toEqual(['High', 'Low']);
+    expect(refusal(container)).toBeNull();
+    expect(container.querySelector('svg')).not.toBeNull();
+    expect(warn).not.toHaveBeenCalled();
+    warn.mockRestore();
+  });
+
+  it('draws what projects when the group key is only PARTIALLY projected', () => {
+    const warn = vi.spyOn(console, 'warn').mockImplementation(() => {});
+    const { container, series, data } = renderPivot([
+      { status: 'Backlog', priority: 'High', est_hours: 12 },
+      { status: 'Done', est_hours: 30 },
+    ]);
+
+    // Premise C, decided by MIRRORING the first dimension rather than by taste:
+    // `hasNoCategoryKey` refuses only when NOT ONE row carries the key
+    // (`!rows.some(…)`), so a partially projected axis draws what projects. The
+    // series axis inherits that line exactly — one row carries `priority`, so
+    // that group has a bucket, `series` is non-empty and nothing refuses.
+    expect(series.map((s) => s.dataKey)).toEqual(['High']);
+    // The unprojected row still contributes its CATEGORY and no measure — the
+    // objectui#4673 rule this card does not touch.
+    expect(data).toEqual([{ status: 'Backlog', High: 12 }, { status: 'Done' }]);
+    expect(refusal(container)).toBeNull();
+    expect(container.querySelectorAll('.recharts-rectangle')).toHaveLength(1);
+    expect(warn).not.toHaveBeenCalled();
+    warn.mockRestore();
+  });
+
+  it('never fires on a single-dimension chart, which has a measure series', () => {
+    const warn = vi.spyOn(console, 'warn').mockImplementation(() => {});
+    const { data, xAxisKey, series } = buildChartSeries(
+      [{ status: 'Backlog', est_hours: 12 }, { status: 'Done', est_hours: 30 }],
+      ['status'],
+      ['est_hours'],
+    );
+    const { container } = render(
+      <AdvancedChartImpl
+        chartType="bar"
+        data={data as any}
+        xAxisKey={xAxisKey}
+        series={series as any}
+        isAnimationActive={false}
+      />,
+    );
+
+    expect(series.map((s) => s.dataKey)).toEqual(['est_hours']);
+    expect(refusal(container)).toBeNull();
+    expect(container.querySelectorAll('.recharts-rectangle')).toHaveLength(2);
+    expect(warn).not.toHaveBeenCalled();
+    warn.mockRestore();
+  });
+
+  it('says nothing about a chart with no rows to plot at all', () => {
+    const warn = vi.spyOn(console, 'warn').mockImplementation(() => {});
+    // Empty is not unreadable — an empty result set is a legitimate answer and
+    // the chart's own empty state owns it. Same rows.length > 0 precondition
+    // the framework#4033 guard carries.
+    const { container } = render(
+      <AdvancedChartImpl chartType="bar" data={[]} xAxisKey="status" series={[]} isAnimationActive={false} />,
+    );
+    expect(refusal(container)).toBeNull();
+    expect(warn).not.toHaveBeenCalled();
+    warn.mockRestore();
+  });
+});

